File: src/main/java/com/atsuishio/superbwarfare/network/message/receive/ClientMotionSyncMessage.java
package com.atsuishio.superbwarfare.network.message.receive;

import com.atsuishio.superbwarfare.network.ClientPacketHandler;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.phys.Vec3;

public class ClientMotionSyncMessage {

    public final int id;
    public final float x;
    public final float y;
    public final float z;

    public ClientMotionSyncMessage(Entity entity) {
        this(entity.getId(), entity.getDeltaMovement());
    }

    public ClientMotionSyncMessage(int id, Vec3 motion) {
        this.id = id;
        this.x = (float) motion.x;
        this.y = (float) motion.y;
        this.z = (float) motion.z;
    }

    public static void encode(ClientMotionSyncMessage message, FriendlyByteBuf buffer) {
        buffer.writeVarInt(message.id);
        buffer.writeFloat(message.x);
        buffer.writeFloat(message.y);
        buffer.writeFloat(message.z);
    }

    public static ClientMotionSyncMessage decode(FriendlyByteBuf buffer) {
        int id = buffer.readVarInt();
        double x = buffer.readFloat();
        double y = buffer.readFloat();
        double z = buffer.readFloat();
        return new ClientMotionSyncMessage(id, new Vec3(x, y, z));
    }

    public static void handler(ClientMotionSyncMessage message) {
        ClientPacketHandler.handleClientSyncMotion(message);
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/network/message/receive/ClientMotionSyncMessage.kt
package com.atsuishio.superbwarfare.network.message.receive

import com.atsuishio.superbwarfare.ksp.annotation.RegisterPacket
import com.atsuishio.superbwarfare.network.ClientPacketPayload
import com.atsuishio.superbwarfare.network.PayloadContext
import com.atsuishio.superbwarfare.entity.projectile.IBulletProperties
import com.atsuishio.superbwarfare.tools.clientLevel
import kotlinx.serialization.Serializable
import net.minecraft.world.entity.Entity
import net.minecraft.world.phys.Vec3

@Serializable
@RegisterPacket
data class ClientMotionSyncMessage(
    val id: Int,
    val x: Float,
    val y: Float,
    val z: Float,
    val explosionRadius: Float? = null,
) : ClientPacketPayload() {

    constructor(id: Int, motion: Vec3) : this(id, motion.x.toFloat(), motion.y.toFloat(), motion.z.toFloat())
    constructor(entity: Entity) : this(
        entity.id,
        entity.deltaMovement.x.toFloat(),
        entity.deltaMovement.y.toFloat(),
        entity.deltaMovement.z.toFloat(),
        (entity as? IBulletProperties)?.getExplosionRadius()
    )

    override fun PayloadContext.handler() {
        val entity = clientLevel?.getEntity(id) ?: return
        entity.lerpMotion(x.toDouble(), y.toDouble(), z.toDouble())
        explosionRadius?.let { (entity as? IBulletProperties)?.setExplosionRadius(it) }
    }
}
